Implement getting node's pre-image based on block height

linked0 opened the pull request:
This PR is for a function getting a pre-image at a certain height from a node. As discussed with @Geod24  in PR #530 , the information for a pre-image(a.k.a. PreimageInfo) becomes to have the block height at which it exists not the round number.

- Make the cycle length value constant
- Add the PreimageInfo structure
- Add the getPreimage function used to get a node's pre-image at a certain block height.
- Unittest

#492

Geod24 commented on source/agora/consensus/data/PreimageInfo.d:
```diff
@@ -0,0 +1,34 @@
+
+/*******************************************************************************
+
+    Define preimage data for a pre-image at a certain height and the height number
```
That definition is a bit redundant don't you think ? (And over 80 chars)

Geod24 commented on source/agora/consensus/data/PreimageInfo.d:
```diff
@@ -0,0 +1,34 @@
+
+public struct PreimageInfo
+{
+    // The key for the enrollment
```
```suggestion
    /// The key for the enrollment
```

Two slashes is a comment, three is a documentation comment

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -73,5 +73,5 @@ unittest
     // check if nodes contains enrollment data previously sent.
     nodes.each!(node =>
-        retryFor(node.hasEnrollment(enroll.utxo_key) == true, 5.seconds));
+        retryFor(node.hasEnrollment(enroll.utxo_key) == true, 10.seconds));
```
Why?

linked0 replied:
> retryFor(node.hasEnrollment(enroll.utxo_key) == true, 10.seconds));
> Why?

It's changed temporally before requesting a review in order to find the reason why the timeout occured in the Github build.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -49,6 +50,9 @@ public class EnrollmentManager
     private Pair key_pair;
 
+    /// Key used for enrollment which is actually an UTXO hash
+    private Hash enroll_key;
```
If it's "actually an UTXO hash", why is it named key then ?

linked0 replied:
> If it's "actually an UTXO hash", why is it named key then ?

Because It's used in order to select an enrollment information by the key, or `enroll_key` in the validator set table. What I intended to represent by the name is that it's the key for an enrollment information.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -493,6 +536,32 @@ public class EnrollmentManager
+
+    public bool getPreimage (ulong height, out PreimageInfo preimage) @safe
+    {
+        auto start_height = this.getEnrolledHeight(this.enroll_key) + 1;
```
```suggestion
        const start_height = this.getEnrolledHeight(this.enroll_key) + 1;
```

Geod24 reviewed (commented):
I guess you also want to add the API entry here ?

linked0 commented:
Ready for review.
It's a part of the PR #530 . 

There are some test error on the linux and dmd-2.090.0 config but I think it's not related this PR. The error messages as follows with some modification on the retry time.
```
2020-02-12 04:18:35,869 Trace [SCP] - NominationProtocol::nominate (1) e68a4f
2020-02-12 04:18:35,869 Trace [SCP] - NominationProtocol::nominate (TIMED OUT)
======================================================================

Module tests failed: agora.test.EnrollmentManager
core.exception.AssertError@source/agora/test/EnrollmentManager.d(76): Check condition failed after timeout of 10 secs and 100 attempts
----------------
Program exited with code 1
```
I will check it anyway.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -316,9 +330,10 @@ public class EnrollmentManager
 
         // K, frozen UTXO hash
         this.data.utxo_key = frozen_utxo_hash;
+        this.enroll_key = frozen_utxo_hash;
```
So I don't see why we need `enroll_key` ?

linked0 replied:
> So I don't see why we need enroll_key ?

It's necessary for selecting a `enrolled_height` from the validator set table. The `enrolled_height` is set in the middle of adding a block. The function that updates the `enrolled_height` is the `addValidatedBlock` which is modified in Henry's PR, or PR #504 . The code is as follows.

```
foreach (enrollment; block.header.enrollments)
    if (!this.enroll_man.updateEnrolledHeight(enrollment.utxo_key,
        block.header.height))
        assert(0);
```

Geod24 commented:
Looking at [travis](https://travis-ci.com/bpfkorea/agora/jobs/286197713) might also be helpful.

```
Running ./build/agora-unittests 
Module tests failed: agora.consensus.EnrollmentManager
core.exception.AssertError@source/agora/consensus/EnrollmentManager.d(670): unittest failure
```

Looks like it's being triggered because one of the tests uses a hash not in the UTXO.

linked0 committed:
Make the variable of validation cycle length constant

The validation cycle length can be used out of the class.
That's why this commit is made.

linked0 committed:
Add the function getting pre-image at a certain height

The function is used to get a pre-image before revealing it into the network.
In order to get a pre-image, a caller must know the key for enrollment
because the enrolled height used to calculate the start height is stored in the
validator set table having the key as a PRIMARY KEY.

linked0 commented:
Ready for review.
I applied all the suggestions and comments to the code and all tests passed. Thank you.
